src/Admin/AbstractAdmin.php

```diff
@@ -55,17 +44,16 @@
 use Symfony\Component\Routing\Generator\UrlGeneratorInterface as RoutingUrlGeneratorInterface;
 use Symfony\Component\Security\Acl\Model\DomainObjectInterface;
 use Symfony\Component\Security\Core\Exception\AccessDeniedException;
-use Symfony\Component\Translation\TranslatorInterface;
 use Symfony\Component\Validator\Mapping\GenericMetadata;
-use Symfony\Component\Validator\Validator\ValidatorInterface;
 
 /**
  * @author Thomas Rabaix <thomas.rabaix@sonata-project.org>
  *
  * @phpstan-template T of object
+ * @phpstan-extends AbstractTaggedAdmin<T>
  * @phpstan-implements AdminInterface<T>
  */
-abstract class AbstractAdmin implements AdminInterface, DomainObjectInterface, AdminTreeInterface
+abstract class AbstractAdmin extends AbstractTaggedAdmin implements AdminInterface, DomainObjectInterface, AdminTreeInterface
```

Comment:
This looks more like a trait to me. We should not overuse class extension

Reply:
I dont think so.

A Trait can't have a construct @core23

And a class with the sonata.admin MUST have the following signature for the first constructor arguments.

Reply:
👍 That's a good indicator for a design issue (that we have for a long time...). If we need a specific signature for the constructor, we should not use it and create a new `initSomething` method for this.

Reply:
Maybe we can use the `initialize` method for this

Reply:
> 👍 That's a good indicator for a design issue (that we have for a long time...). If we need a specific signature for the constructor, we should not use it and create a new `initSomething` method for this.

When defining an admin, we're writing 
```
    admin.category:
        class: App\Admin\CategoryAdmin
        arguments: [~, App\Entity\Category, ~]
        tags:
            - { name: sonata.admin, manager_type: orm, label: Category }
```
The `arguments` key, is by definition the arguments of the `__construct()` method,
That's how service declaration is working for symfony, and that's how it's currently working for Sonata.

If it's a design issue, it's here from the first day. I introduced nothing.
You're just discovering or re-discovering this potential design issue, because of the clean/refacto.
So it shouldn't be a blocking point of this PR. Do you agree @core23 ?

Moreover I'm just trying to cleaning interface before Sonata 4.
The last refacto started is about `TemplateRegistry` and since it's not finished it's blocking the major release.
It's not time for another big refacto IMHO, and should be waiting Sonata 5 then.

Reply:
> Maybe we can use the `initialize` method for this

IMHO, we could just migrate from
```
admin.category:
        class: App\Admin\CategoryAdmin
        arguments: [~, App\Entity\Category, ~]
        tags:
            - { name: sonata.admin, manager_type: orm, label: Category }
```
to
```
admin.category:
        class: App\Admin\CategoryAdmin
        arguments: [App\Entity\Category]
        tags:
            - { name: sonata.admin, manager_type: orm, label: Category }
```

And from 
```
admin.category:
        class: App\Admin\CategoryAdmin
        arguments: [CustomCode, App\Entity\Category, CustomController::class]
        tags:
            - { name: sonata.admin, manager_type: orm, label: Category }
```
to 
```
admin.category:
        class: App\Admin\CategoryAdmin
        arguments: [App\Entity\Category]
        tags:
            - { name: sonata.admin, code: CustomCode, controller:  CustomController::class, manager_type: orm, label: Category }
```

WDYT ?

Reply:
Seems like a good solution to me :+1:

Reply:
@core23 Is it ok for you to delay this in another PR/milestone ?

The design issue is here for a long time like you said and shouldn't be a blocker here. I can't solve everything.
We just have to do small steps. Especially when we wants to release 4.0.